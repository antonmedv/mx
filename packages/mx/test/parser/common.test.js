const test = require('ava')
const parse = require('./parse')

test('parse single tag', t => {
  const code = `
  div
  `
  t.snapshot(parse(code))
})

test('parse indent and dedent', t => {
  const code = `
  body
    header
      h1
    section
      div
      p
    footer
  `
  t.snapshot(parse(code))
})

test('ignore empty lines', t => {
  const code = `
  body
    div
` + '\n' +
`      h1
` + '      \n' +
`    footer
  `
  t.snapshot(parse(code))
})

test('parse attributes', t => {
  const code = `
  div id="code" class="cool"
  `
  t.snapshot(parse(code))
})

test('parse attributes with children', t => {
  const code = `
  div id="code"
    p
  `
  t.snapshot(parse(code))
})

test('parse tag with content', t => {
  const code = `
  p Beware of dragons ðŸ”¥
  `
  t.snapshot(parse(code))
})

test('parse attributes with content', t => {
  const code = `
  div class="hi" Hello world!
  `
  t.snapshot(parse(code))
})

test('parse tag with block of text', t => {
  const code = `
  div
    p.
      Beware of dragons

      Dragons are very dangerous creatures,
      They live in a caves in the mountains.
    footer
  `
  t.snapshot(parse(code))
})

test('parse complex selector', t => {
  const code = `
  #id.class-name
  `
  t.snapshot(parse(code))
})

test('parse complex selector with dot on end', t => {
  const code = `
  .intro.
    Please don't use this king of syntax ðŸ˜‰
  `
  t.snapshot(parse(code))
})

test('parse text tags', t => {
  const code = `
  p
    | Here goes some text
  `
  t.snapshot(parse(code))
})
