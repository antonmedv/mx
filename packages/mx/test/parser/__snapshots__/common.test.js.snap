// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ignore empty lines 1`] = `
DocumentNode {
  "body": Array [
    ElementNode {
      "attributes": Array [],
      "body": Array [
        ElementNode {
          "attributes": Array [],
          "body": Array [
            ElementNode {
              "attributes": Array [],
              "body": Array [],
              "loc": SourceLocation {
                "end": Position {
                  "column": 0,
                  "line": 5,
                },
                "source": "file.mx",
                "start": Position {
                  "column": 4,
                  "line": 4,
                },
              },
              "name": "h1",
              "type": "Element",
            },
          ],
          "loc": SourceLocation {
            "end": Position {
              "column": 2,
              "line": 6,
            },
            "source": "file.mx",
            "start": Position {
              "column": 2,
              "line": 2,
            },
          },
          "name": "div",
          "type": "Element",
        },
        ElementNode {
          "attributes": Array [],
          "body": Array [],
          "loc": SourceLocation {
            "end": Position {
              "column": 0,
              "line": 7,
            },
            "source": "file.mx",
            "start": Position {
              "column": 2,
              "line": 6,
            },
          },
          "name": "footer",
          "type": "Element",
        },
      ],
      "loc": SourceLocation {
        "end": Position {
          "column": 0,
          "line": 7,
        },
        "source": "file.mx",
        "start": Position {
          "column": 0,
          "line": 1,
        },
      },
      "name": "body",
      "type": "Element",
    },
  ],
  "loc": SourceLocation {
    "end": Position {
      "column": 0,
      "line": 7,
    },
    "source": "file.mx",
    "start": Position {
      "column": 0,
      "line": 1,
    },
  },
  "type": "Document",
}
`;

exports[`parse attributes 1`] = `
DocumentNode {
  "body": Array [
    ElementNode {
      "attributes": Array [
        AttributeNode {
          "body": Array [
            LiteralNode {
              "loc": SourceLocation {
                "end": Position {
                  "column": 12,
                  "line": 1,
                },
                "source": "file.mx",
                "start": Position {
                  "column": 8,
                  "line": 1,
                },
              },
              "type": "Literal",
              "value": "\\"code\\"",
            },
          ],
          "loc": SourceLocation {
            "end": Position {
              "column": 13,
              "line": 1,
            },
            "source": "file.mx",
            "start": Position {
              "column": 4,
              "line": 1,
            },
          },
          "name": "id",
          "type": "Attribute",
        },
        AttributeNode {
          "body": Array [
            LiteralNode {
              "loc": SourceLocation {
                "end": Position {
                  "column": 25,
                  "line": 1,
                },
                "source": "file.mx",
                "start": Position {
                  "column": 21,
                  "line": 1,
                },
              },
              "type": "Literal",
              "value": "\\"cool\\"",
            },
          ],
          "loc": SourceLocation {
            "end": Position {
              "column": 26,
              "line": 1,
            },
            "source": "file.mx",
            "start": Position {
              "column": 14,
              "line": 1,
            },
          },
          "name": "class",
          "type": "Attribute",
        },
      ],
      "body": Array [],
      "loc": SourceLocation {
        "end": Position {
          "column": 0,
          "line": 2,
        },
        "source": "file.mx",
        "start": Position {
          "column": 0,
          "line": 1,
        },
      },
      "name": "div",
      "type": "Element",
    },
  ],
  "loc": SourceLocation {
    "end": Position {
      "column": 0,
      "line": 2,
    },
    "source": "file.mx",
    "start": Position {
      "column": 0,
      "line": 1,
    },
  },
  "type": "Document",
}
`;

exports[`parse attributes with children 1`] = `
DocumentNode {
  "body": Array [
    ElementNode {
      "attributes": Array [
        AttributeNode {
          "body": Array [
            LiteralNode {
              "loc": SourceLocation {
                "end": Position {
                  "column": 12,
                  "line": 1,
                },
                "source": "file.mx",
                "start": Position {
                  "column": 8,
                  "line": 1,
                },
              },
              "type": "Literal",
              "value": "\\"code\\"",
            },
          ],
          "loc": SourceLocation {
            "end": Position {
              "column": 13,
              "line": 1,
            },
            "source": "file.mx",
            "start": Position {
              "column": 4,
              "line": 1,
            },
          },
          "name": "id",
          "type": "Attribute",
        },
      ],
      "body": Array [
        ElementNode {
          "attributes": Array [],
          "body": Array [],
          "loc": SourceLocation {
            "end": Position {
              "column": 0,
              "line": 3,
            },
            "source": "file.mx",
            "start": Position {
              "column": 2,
              "line": 2,
            },
          },
          "name": "p",
          "type": "Element",
        },
      ],
      "loc": SourceLocation {
        "end": Position {
          "column": 0,
          "line": 3,
        },
        "source": "file.mx",
        "start": Position {
          "column": 0,
          "line": 1,
        },
      },
      "name": "div",
      "type": "Element",
    },
  ],
  "loc": SourceLocation {
    "end": Position {
      "column": 0,
      "line": 3,
    },
    "source": "file.mx",
    "start": Position {
      "column": 0,
      "line": 1,
    },
  },
  "type": "Document",
}
`;

exports[`parse attributes with content 1`] = `
DocumentNode {
  "body": Array [
    ElementNode {
      "attributes": Array [
        AttributeNode {
          "body": Array [
            LiteralNode {
              "loc": SourceLocation {
                "end": Position {
                  "column": 13,
                  "line": 1,
                },
                "source": "file.mx",
                "start": Position {
                  "column": 11,
                  "line": 1,
                },
              },
              "type": "Literal",
              "value": "\\"hi\\"",
            },
          ],
          "loc": SourceLocation {
            "end": Position {
              "column": 14,
              "line": 1,
            },
            "source": "file.mx",
            "start": Position {
              "column": 4,
              "line": 1,
            },
          },
          "name": "class",
          "type": "Attribute",
        },
      ],
      "body": Array [
        TextNode {
          "loc": SourceLocation {
            "end": Position {
              "column": 20,
              "line": 1,
            },
            "source": "file.mx",
            "start": Position {
              "column": 15,
              "line": 1,
            },
          },
          "text": "Hello",
          "type": "Text",
        },
        TextNode {
          "loc": SourceLocation {
            "end": Position {
              "column": 26,
              "line": 1,
            },
            "source": "file.mx",
            "start": Position {
              "column": 21,
              "line": 1,
            },
          },
          "text": "world",
          "type": "Text",
        },
        TextNode {
          "loc": SourceLocation {
            "end": Position {
              "column": 27,
              "line": 1,
            },
            "source": "file.mx",
            "start": Position {
              "column": 26,
              "line": 1,
            },
          },
          "text": "!",
          "type": "Text",
        },
      ],
      "loc": SourceLocation {
        "end": Position {
          "column": 0,
          "line": 2,
        },
        "source": "file.mx",
        "start": Position {
          "column": 0,
          "line": 1,
        },
      },
      "name": "div",
      "type": "Element",
    },
  ],
  "loc": SourceLocation {
    "end": Position {
      "column": 0,
      "line": 2,
    },
    "source": "file.mx",
    "start": Position {
      "column": 0,
      "line": 1,
    },
  },
  "type": "Document",
}
`;

exports[`parse complex selector 1`] = `
DocumentNode {
  "body": Array [
    ElementNode {
      "attributes": Array [],
      "body": Array [],
      "loc": SourceLocation {
        "end": Position {
          "column": 0,
          "line": 2,
        },
        "source": "file.mx",
        "start": Position {
          "column": 0,
          "line": 1,
        },
      },
      "name": "#id.class-name",
      "type": "Element",
    },
  ],
  "loc": SourceLocation {
    "end": Position {
      "column": 0,
      "line": 2,
    },
    "source": "file.mx",
    "start": Position {
      "column": 0,
      "line": 1,
    },
  },
  "type": "Document",
}
`;

exports[`parse complex selector with dot on end 1`] = `
DocumentNode {
  "body": Array [
    ElementNode {
      "attributes": Array [],
      "body": Array [
        TextNode {
          "loc": SourceLocation {
            "end": Position {
              "column": 0,
              "line": 3,
            },
            "source": "file.mx",
            "start": Position {
              "column": 0,
              "line": 2,
            },
          },
          "text": "  Please don't use this king of syntax 😉
",
          "type": "Text",
        },
      ],
      "loc": SourceLocation {
        "end": Position {
          "column": 0,
          "line": 3,
        },
        "source": "file.mx",
        "start": Position {
          "column": 0,
          "line": 1,
        },
      },
      "name": ".intro",
      "type": "Element",
    },
  ],
  "loc": SourceLocation {
    "end": Position {
      "column": 0,
      "line": 3,
    },
    "source": "file.mx",
    "start": Position {
      "column": 0,
      "line": 1,
    },
  },
  "type": "Document",
}
`;

exports[`parse indent and dedent 1`] = `
DocumentNode {
  "body": Array [
    ElementNode {
      "attributes": Array [],
      "body": Array [
        ElementNode {
          "attributes": Array [],
          "body": Array [
            ElementNode {
              "attributes": Array [],
              "body": Array [],
              "loc": SourceLocation {
                "end": Position {
                  "column": 0,
                  "line": 4,
                },
                "source": "file.mx",
                "start": Position {
                  "column": 4,
                  "line": 3,
                },
              },
              "name": "h1",
              "type": "Element",
            },
          ],
          "loc": SourceLocation {
            "end": Position {
              "column": 2,
              "line": 4,
            },
            "source": "file.mx",
            "start": Position {
              "column": 2,
              "line": 2,
            },
          },
          "name": "header",
          "type": "Element",
        },
        ElementNode {
          "attributes": Array [],
          "body": Array [
            ElementNode {
              "attributes": Array [],
              "body": Array [],
              "loc": SourceLocation {
                "end": Position {
                  "column": 0,
                  "line": 6,
                },
                "source": "file.mx",
                "start": Position {
                  "column": 4,
                  "line": 5,
                },
              },
              "name": "div",
              "type": "Element",
            },
            ElementNode {
              "attributes": Array [],
              "body": Array [],
              "loc": SourceLocation {
                "end": Position {
                  "column": 0,
                  "line": 7,
                },
                "source": "file.mx",
                "start": Position {
                  "column": 4,
                  "line": 6,
                },
              },
              "name": "p",
              "type": "Element",
            },
          ],
          "loc": SourceLocation {
            "end": Position {
              "column": 2,
              "line": 7,
            },
            "source": "file.mx",
            "start": Position {
              "column": 2,
              "line": 4,
            },
          },
          "name": "section",
          "type": "Element",
        },
        ElementNode {
          "attributes": Array [],
          "body": Array [],
          "loc": SourceLocation {
            "end": Position {
              "column": 0,
              "line": 8,
            },
            "source": "file.mx",
            "start": Position {
              "column": 2,
              "line": 7,
            },
          },
          "name": "footer",
          "type": "Element",
        },
      ],
      "loc": SourceLocation {
        "end": Position {
          "column": 0,
          "line": 8,
        },
        "source": "file.mx",
        "start": Position {
          "column": 0,
          "line": 1,
        },
      },
      "name": "body",
      "type": "Element",
    },
  ],
  "loc": SourceLocation {
    "end": Position {
      "column": 0,
      "line": 8,
    },
    "source": "file.mx",
    "start": Position {
      "column": 0,
      "line": 1,
    },
  },
  "type": "Document",
}
`;

exports[`parse single tag 1`] = `
DocumentNode {
  "body": Array [
    ElementNode {
      "attributes": Array [],
      "body": Array [],
      "loc": SourceLocation {
        "end": Position {
          "column": 0,
          "line": 2,
        },
        "source": "file.mx",
        "start": Position {
          "column": 0,
          "line": 1,
        },
      },
      "name": "div",
      "type": "Element",
    },
  ],
  "loc": SourceLocation {
    "end": Position {
      "column": 0,
      "line": 2,
    },
    "source": "file.mx",
    "start": Position {
      "column": 0,
      "line": 1,
    },
  },
  "type": "Document",
}
`;

exports[`parse single tag 2`] = `
DocumentNode {
  "body": Array [
    TagNode {
      "attributes": Array [],
      "body": Array [],
      "loc": SourceLocation {
        "end": Position {
          "column": 3,
          "line": 1,
        },
        "source": "file.mx",
        "start": Position {
          "column": 0,
          "line": 1,
        },
      },
      "name": "div",
      "type": "tag",
    },
  ],
  "loc": SourceLocation {
    "end": Position {
      "column": 3,
      "line": 1,
    },
    "source": "file.mx",
    "start": Position {
      "column": 0,
      "line": 1,
    },
  },
  "type": "Document",
}
`;

exports[`parse tag with block of text 1`] = `
DocumentNode {
  "body": Array [
    ElementNode {
      "attributes": Array [],
      "body": Array [
        ElementNode {
          "attributes": Array [],
          "body": Array [
            TextNode {
              "loc": SourceLocation {
                "end": Position {
                  "column": 0,
                  "line": 7,
                },
                "source": "file.mx",
                "start": Position {
                  "column": 0,
                  "line": 3,
                },
              },
              "text": "    Beware of dragons

    Dragons are very dangerous creatures,
    They live in a caves in the mountains.
",
              "type": "Text",
            },
          ],
          "loc": SourceLocation {
            "end": Position {
              "column": 0,
              "line": 7,
            },
            "source": "file.mx",
            "start": Position {
              "column": 2,
              "line": 2,
            },
          },
          "name": "p",
          "type": "Element",
        },
        ElementNode {
          "attributes": Array [],
          "body": Array [],
          "loc": SourceLocation {
            "end": Position {
              "column": 0,
              "line": 8,
            },
            "source": "file.mx",
            "start": Position {
              "column": 2,
              "line": 7,
            },
          },
          "name": "footer",
          "type": "Element",
        },
      ],
      "loc": SourceLocation {
        "end": Position {
          "column": 0,
          "line": 8,
        },
        "source": "file.mx",
        "start": Position {
          "column": 0,
          "line": 1,
        },
      },
      "name": "div",
      "type": "Element",
    },
  ],
  "loc": SourceLocation {
    "end": Position {
      "column": 0,
      "line": 8,
    },
    "source": "file.mx",
    "start": Position {
      "column": 0,
      "line": 1,
    },
  },
  "type": "Document",
}
`;

exports[`parse tag with content 1`] = `
DocumentNode {
  "body": Array [
    ElementNode {
      "attributes": Array [],
      "body": Array [
        TextNode {
          "loc": SourceLocation {
            "end": Position {
              "column": 8,
              "line": 1,
            },
            "source": "file.mx",
            "start": Position {
              "column": 2,
              "line": 1,
            },
          },
          "text": "Beware",
          "type": "Text",
        },
        TextNode {
          "loc": SourceLocation {
            "end": Position {
              "column": 11,
              "line": 1,
            },
            "source": "file.mx",
            "start": Position {
              "column": 9,
              "line": 1,
            },
          },
          "text": "of",
          "type": "Text",
        },
        TextNode {
          "loc": SourceLocation {
            "end": Position {
              "column": 19,
              "line": 1,
            },
            "source": "file.mx",
            "start": Position {
              "column": 12,
              "line": 1,
            },
          },
          "text": "dragons",
          "type": "Text",
        },
        TextNode {
          "loc": SourceLocation {
            "end": Position {
              "column": 22,
              "line": 1,
            },
            "source": "file.mx",
            "start": Position {
              "column": 20,
              "line": 1,
            },
          },
          "text": "🔥",
          "type": "Text",
        },
      ],
      "loc": SourceLocation {
        "end": Position {
          "column": 0,
          "line": 2,
        },
        "source": "file.mx",
        "start": Position {
          "column": 0,
          "line": 1,
        },
      },
      "name": "p",
      "type": "Element",
    },
  ],
  "loc": SourceLocation {
    "end": Position {
      "column": 0,
      "line": 2,
    },
    "source": "file.mx",
    "start": Position {
      "column": 0,
      "line": 1,
    },
  },
  "type": "Document",
}
`;
